L1Trigger/L1THGCal/src/concentrator/HGCalConcentratorSuperTriggerCellImpl.cc

```diff
@@ -86,53 +123,310 @@ HGCalConcentratorSuperTriggerCellImpl::getSuperTriggerCellId(int detid) const {
+    }
+  }
+
+  for (const l1t::HGCalTriggerCell & tc : trigCellVecInput) {
+     if (tc.subdetId() == HGCHEB) continue;
+
+     int thickness = triggerTools_.thicknessIndex(tc.detId(),true);
+     if ( thickness > 0 ){
+
+      if (!(tc.detId() & 1)) { //if even
+        const auto & ctc = coarseTCs[getSuperTriggerCellId(tc.detId(),2)]; 
+        auto & stc = STCs[getSuperTriggerCellId(tc.detId())]; 
+        trigCellVecOutput.push_back( tc );
+        ctc.assignEnergy(trigCellVecOutput.back(), "STC");      
+        //        //reassign max id to the correct one;
+        if ( trigCellVecOutput.back().hwPt() >= stc.GetMaxHwPt() ){
+
+          if ( stc.GetMaxId() != tc.detId() ){
+            stc.SetMaxTC( trigCellVecOutput.back() );
+          }
+
+        }
+      }
+
+    }
+    else{
+
+      trigCellVecOutput.push_back( tc );
+
+    }
+
+
+  }
+
+
+}
+
+
 void 
 HGCalConcentratorSuperTriggerCellImpl::
 superTriggerCellSelectImpl(const std::vector<l1t::HGCalTriggerCell>& trigCellVecInput, std::vector<l1t::HGCalTriggerCell>& trigCellVecOutput)
 { 
 
+  fixedDataSize_ = true;
+
   std::unordered_map<unsigned,SuperTriggerCell> STCs;
+  std::vector<l1t::HGCalTriggerCell> trigCellVecInputEnlarged;
 
-  // first pass, fill the super trigger cells
+  // first pass, fill the "coarse" trigger cells
   for (const l1t::HGCalTriggerCell & tc : trigCellVecInput) {
+    trigCellVecInputEnlarged.push_back ( tc );
     if (tc.subdetId() == HGCHEB) continue;
     STCs[getSuperTriggerCellId(tc.detId())].add(tc);
   }
-    
+
+
+  //The missing trigger cells are needed to be created both for coarsening the existing TC
+  // in the thick modules (for the fixed data size choice), or for any of the energy spread algorithms (except super TCs)
+
+  if ( fixedDataSize_ == true || energyDivisionType_!=superTriggerCell){
+    createMissingTriggerCells( STCs, trigCellVecInputEnlarged);
+  }
+
+
+  //Coarsen if needed
+  std::vector<l1t::HGCalTriggerCell> trigCellVecInputCoarsened;
+  if ( fixedDataSize_ == true ){
+    coarsenTriggerCells( STCs, trigCellVecInputEnlarged, trigCellVecInputCoarsened);
+  }
+  else{
+    trigCellVecInputCoarsened = trigCellVecInputEnlarged;
+  }
+
+  
+  if ( energyDivisionType_ == oneBitFraction ){
+    //Get the 1 bit fractions. There should be exactly 4 TCs per STC
+    for (const l1t::HGCalTriggerCell & tc : trigCellVecInputCoarsened) {
+      if (tc.subdetId() == HGCHEB) continue;
+      STCs[getSuperTriggerCellId(tc.detId())].getFractionSum(tc);
+    }
+  }
+
   // second pass, write them out
-  for (const l1t::HGCalTriggerCell & tc : trigCellVecInput) {
-    
+  for (const l1t::HGCalTriggerCell & tc : trigCellVecInputCoarsened) {
     //If scintillator use a simple threshold cut
     if (tc.subdetId() == HGCHEB) {
       trigCellVecOutput.push_back( tc );
     } else {
       const auto & stc = STCs[getSuperTriggerCellId(tc.detId())]; 
-      if (tc.detId() == stc.GetMaxId() ) {
+      if ( stc.rejected() ) continue;
```

Comment:
Why would a STC be rejected?

Reply:
Sometimes a newly created TC doesn't exist in the geometry (at the edges). The values are set to NaN. In this case I reject the whole STC (conservatively).

Reply:
I think it would be better to still have the STC, even though part of its constituents is missing.